TODO/Breaking-Swift-with-reference-counted-structs.md

```diff
@@ -1,303 +1,283 @@
     }
+}
+```
 
+编译器在 Swift 2.2 报了一个「内存区段错误」(segmentation fault)，而在 Swift 开发版本快照 (Swift Development Snapshot) 2016-03-26 版本则报了一个「致命错误」(fatal error)。
 
+"Excellent!"，我现在很开心(I'm Angry!)。
 
-The compiler throws a segmentation fault in Swift 2.2 and a fatal error in Swift Development Snapshot 2016-03-24.
-
-Excellent! I’m having fun already.
-
-Of course, I could avoid all compiler problems by doing this:
-
-
+当然，我会这样做来避免编译器报错：
 
-    struct Counter {
-        var count: Int
-        let od: OnDelete
-        init() {
-            let c = 0
-            count = c
-            od = OnDelete { print("Counter value is \(c)") }
-        }
+```swift
+struct Counter {
+    var count: Int
+    let od: OnDelete
+    init() {
+        let c = 0
+        count = c
+        od = OnDelete { print("Counter value is \(c)") }
     }
+}
+```
 
+或者用另一种不常见的方法，在这种情况下它们是等效的：
 
-
-or the seldom-seen capture list which, in this case, is equivalent:
-
-
-
-    struct Counter {
-        var count = 0
-        let od: OnDelete?
-        init() {
-            od = OnDelete { [count] in print("Counter value is \(count)") }
-        }
+```swift
+struct Counter {
+    var count = 0
+    let od: OnDelete?
+    init() {
+        od = OnDelete { [count] in print("Counter value is \(count)") }
     }
+}
+```
 
+可是这两个方法并不能真的让我们访问到这个结构体本身。因为这两种方法捕捉到的都只是 `count` 的不可变副本，但是我们想要得到的是最新的 `count` 可变值。
 
-
-but neither of these options actually let us access the `struct` itself; both these options capture an immutable copy of the `count` field but we want access to the up-to-date mutable `count`.
-
-
-
-    struct Counter {
-        var count = 0
-        var od: OnDelete?
-        init() {
-            od = OnDelete { print("Counter value is \(self.count)") }
-        }
+```swift
+struct Counter {
+    var count = 0
+    var od: OnDelete?
+    init() {
+        od = OnDelete { print("Counter value is \(self.count)") }
     }
+}
+```
 
+万岁！这一切都是可变的并且共享的，我们已经成功的捕获了这个变量，编译器也没有崩溃。
 
+我们应该来尝试使用这个代码，因为他能很好的工作，不是吗？
 
-Hooray! That’s better. Everything is mutable and shared. We’ve captured the `count` variable and there are no compiler crashes.
-
-We should ship this code since it clearly works, doesn’t it?
-
-## Completely loopy
+## 疯狂的循环 <a name="completely-loopy" />
 
-It clearly doesn’t work. If we run the code the same way as before:
-
-
-
-           do {
-        let c = Counter()
-        print("Not deleted, yet")
-        withExtendedLifetime(c) {}
-    }
+如果我们运行代码的方式和一起一样的话，显然是不行的：
 
+```swift
+do {
+    let c = Counter()
+    print("Not deleted, yet")
+    withExtendedLifetime(c) {}
+}
+```
 
+我们只会得到这样的输出：
 
-the only output we get is:
+```bash
+Not deleted, yet
+```
 
-    Not deleted, yet
+这个 `OnDelete` 闭包没有被调用，为什么？
 
-The `OnDelete` closure is not getting invoked. Why?
+通过查看 SIL(Swift Intermediate Language，Swift 中继语言，通过 `swiftc -emit-sil` 命令返回)，很显然在 `OnDelete` 的闭包里阻止了 `self` 被优化到堆中。这就意味着使用 `alloc_stack` 被 `alloc_box` 替代用来分配 `self` 变量：
```

Comment:
原文：This means that instead of using `alloc_stack`, the `self` variable is allocated using `alloc_box`:
建议：这意味着并非`alloc_stack`，`self`变量是通过`alloc_box`分配的